File: Ruby/quick_sort.rb
def quicksort(arr, left = 0, right = nil)
  right = arr.length - 1 if right.nil?

  if left < right
    pivot = right

    partition_index = partition(arr, pivot, left, right)

    quicksort(arr, left, partition_index - 1)
    quicksort(arr, partition_index + 1, right)
  end
  arr
end

def partition(arr, pivot, left, right)
  pivot_value = arr[pivot]
  index = left

  (left..right).each do |i|
    if arr[i] < pivot_value
      arr[i], arr[index] = arr[index], arr[i]
      index += 1
    end
  end
  arr[right], arr[index] = arr[index], arr[right]
  index
end

File: lua/quicksort.lua
function partition(array, left, right, pivotIndex)
	local pivotValue = array[pivotIndex]
	array[pivotIndex], array[right] = array[right], array[pivotIndex]
	
	local storeIndex = left
	
	for i =  left, right-1 do
    	if array[i] <= pivotValue then
	        array[i], array[storeIndex] = array[storeIndex], array[i]
	        storeIndex = storeIndex + 1
		end
		array[storeIndex], array[right] = array[right], array[storeIndex]
	end
	
   return storeIndex
end

function quicksort(array, left, right)
	if right > left then
	    local pivotNewIndex = partition(array, left, right, left)
	    quicksort(array, left, pivotNewIndex - 1)
	    quicksort(array, pivotNewIndex + 1, right)
	end
end
